Lexically sort schema file names to satisfy order dependencies.

normangilmore opened the pull request:
Hi @floraxue @stefanv 

When following the instructions for installing TextThresher on Ubuntu 14.04, I received a Python stacktrace while running init_docker.sh.

Traceback (most recent call last):
(lines removed)
  File "load_data.py", line 124, in load_schema_dir
    load_schema(parse_schema(os.path.join(dirpath, schema_file)))
  File "load_data.py", line 26, in load_schema
    parent = Topic.objects.get(name=schema_parent)
(lines removed)
thresher.models.DoesNotExist: Topic matching query does not exist.

The problem is that schema1.txt was being loaded before schema0.txt in load_data.py, so the parent topic wasn't being found.

I assume os.listdir happens to return lexical order on Mac directories, but on Ubuntu, it is unordered.

Python 2.7.6 (default, Jun 22 2015, 17:58:13)
[GCC 4.8.2] on linux2

> > > import os
> > > os.listdir('data/sample/schema/')
> > > ['schema1.txt', 'schema0.txt']
> > > sorted(os.listdir('data/sample/schema/'))
> > > ['schema0.txt', 'schema1.txt']

When I added the one function call to sort the schema files lexically, the data all loads and everything works.
Note that schema11.txt will load before schema2.txt, so maybe this needs a readme.md about the lexical sort in the schema folder?

normangilmore committed:
Lexically sort schema file names to satisfy order dependencies.

stefanv commented:
There's another PR open that does this, but yes, sorting is required.

normangilmore commented:
@stefanv #5 has the same change as this one, however, it looks like an unrelated change snuck into that commit which conflicts with my #13, so can we merge this one? I cannot merge yet, would you mind merging?

btburton42 commented on load_data.py:
```diff
@@ -118,7 +118,7 @@ def load_article(article):
 
 def load_schema_dir(dirpath):
-    for schema_file in os.listdir(dirpath):
+    for schema_file in sorted(os.listdir(dirpath)):
```
> How about glob.glob('*.txt')

@stefanv is your comment about globbing still relevant? would that make things more flexible?

normangilmore replied:
It would be nice to limit to `*.txt` as then a README.md or similar could be dropped into the directory without interfering with data loading. I can add the `glob.glob('*.txt')`, but that is a nice-to-have and the current patch is 'must-have'.

btburton42 commented:
👍  do it. thanks norman.

normangilmore commented:
@stefanv @btburton42 
I started to add globbing, and now I see the very next line of code is `if os.path.splitext(article_file)[1] != '.txt':`

So the code is already filtering for just text files to import. I do not see a need to change to globbing.

I will merge this PR Monday if no one objects.

stefanv commented:
Globbing seems cleaner, hence the suggestion.

On September 10, 2016 18:11:35 Norman Gilmore notifications@github.com wrote:

> @stefanv @btburton42
> I started to add globbing, and now I see the very next line of code is `if 
> os.path.splitext(article_file)[1] != '.txt':`
> 
> So the code is already filtering for just text files to import. I do not 
> see a need to change to globbing.
> 
> I will merge this PR Monday if no one objects.
> 
> ## 
> 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/Goodly/TextThresher/pull/11#issuecomment-246123410

normangilmore commented:
Since the glob function returns full paths, but the list needs to be sorted by filename, the resulting list needs to be split for sorting.

``` python
In [19]: glob.glob(os.path.join('data/sample/schema', '*.txt'))
Out[19]: ['data/sample/schema/schema1.txt', 'data/sample/schema/schema0.txt']
```

And because globbing adds support for recursive directory traversal, the returned paths have to be preserved. For example:
`glob.glob(os.path.join('data/sample/**', '*.txt')`

In multiple steps it looks like:

``` python
    datafiles1 = glob.glob(os.path.join(dirpath, '*.txt'))
    datafiles2 = [ os.path.split(path) for path in datafiles1]
    datafiles3 = sorted(datafiles2, None, lambda f: f[1])
    datafiles4 = [ os.path.join(path, fname) for path, fname in datafiles3 ]
    for schema_file in datafiles4:
        load_schema(parse_schema(schema_file)))
```

or composed together in one line:

``` python
    datafiles = sorted(glob.glob(os.path.join(dirpath, '*.txt'),
                       None,
                       lambda f: os.path.split(f)[1]
                 )
    for schema_file in datafiles:
        load_schema(parse_schema(schema_file)))
```

I wasn't sure before trying it, but I think the simplest solution is already in the code.

stefanv commented:
@normangilmore The current function doesn't do recursive listing, so what about

```
for schema_file in sorted(glob(pjoin(dirpath, '*.txt'))):
    load_schema(parse_schema(schema_file))
```

All the files have the same base directory, so sorting should work fine.

normangilmore commented:
The dirpath comes from a command line argument, and therefore may be specified to include multiple paths by the caller. In my view, this adds a potential failure mode to save one line of code. I am not seeing this as cleaner.

stefanv commented:
How would multiple paths get passed through from argparse?  I don't
think this can happen, and even if it could the current code does not
support that case either.

There's an advantage to using well known patterns: other programmers
read it, recognize it, and move on.  But if we use an extra filter, the
programmer has to check what happens first.  If there's a clear
advantage to doing so, then sure, but I don't see it here.

Anyway, this is not worth spending any time on, so please feel free to
go ahead with any option you prefer.

normangilmore commented:
If you quote the command line argument and use a globstar, argparse will pass the argument with globstar on into the program, and then glob can return results from multiple paths.
`python load_data.py --schema-dir='data/sample/**'`

I thought it was a good suggestion so I tried it, but in this particular case, I think the need for sorting by filename makes the glob solution either verbose or subject to the original problem of not reliably sorting the files. If we didn't need the sort, I'd be on board.

I do appreciate the code review and quick feedback, thank you.

normangilmore committed:
Use fnmatch to select schema .txt files to import.

normangilmore commented:
@stefanv I think I found a solution that works for both of us - I used fnmatch.filter to remove the filename extension test from the loop, and it's only one line.

stefanv commented:
Looks good, thanks!
